### DISCUSSION TITLE
Make Calls As User on Backend

### DISCUSSION THREAD

**Discussion by MelbourneDeveloper** [supabase-community/supabase-csharp#98] [General]
The Supabase client generally runs on devices so that phones or desktop apps etc. can talk directly to Supabase instead of making API calls.

In this case, I want to run code on the backend with C#. However, I want to run the code as an existing user. This is an API call, and the API accepts the `accessToken` from the app. The user authenticates on the device and sends the `accessToken` to this API endpoint.

I need the user details from the token. How can I force the current user to load here? `CurrentUser` is always null.

Why isn't `SetAuth` working? 

```csharp
var supabase = new Client(
    "[MYURL]",
    ApiToken,
    new SupabaseOptions {  AutoConnectRealtime = true }
);

supabase.Auth.SetAuth(accessToken);
var result = await supabase.InitializeAsync();

if (supabase.Auth.CurrentUser?.Id == null)
{
    return Results.Unauthorized();
}
```

**Comment by acupofjose** [comment]
@MelbourneDeveloper please try using version v0.13.0 - it references the newer version of gotrue-csharp@4.2.1 which adds the `SetSession(string accessToken, string refreshToken)` method. This will correctly set the authorization as you'd expect.

**Reply by MelbourneDeveloper** [reply]
@acupofjose thanks. I'm trying this now, but it forces me to specify the refresh token, which I don't have. How do I get a refresh token? There is a `RefreshToken` method, but I have to start a session before I can call this, and the point is: I can't start a session without the refresh token.

**Comment by wiverson** [comment]
@MelbourneDeveloper also I'll just chime in that you probably should look at the StatelessClient APIs for running on a server. https://github.com/supabase-community/supabase-csharp/blob/master/Supabase/StatelessClient.cs
